spec/locales-currencies-tz.html

```diff
@@ -285,6 +347,46 @@ <h1>
         1. Return _result_.
       </emu-alg>
     </emu-clause>
+
+      <emu-clause id="sec-string-split-to-list" type="abstract operation">
+        <h1>
+          StringSplitToList (
+            _S_: a String,
+            _separator_: a String that is not the empty String,
+            _limit_: a mathematical value in the range 1 to 2<sup>32</sup> - 1
+          ): a List of Strings
+        </h1>
+        <dl class="header">
+          <dt>description</dt>
+          <dd>
+            The returned List contains substrings of _S_.
+            These substrings are determined by searching from left to right for occurrences of _separator_; these occurrences are not part of any String in the returned List, but serve to divide _S_ into substrings.
+            The output List will contain no more than _limit_ elements; any additional separators and/or substrings present in _S_ will be ignored.
+          </dd>
+        </dl>
+        <emu-alg>
+          1. If _S_ is the empty String, return « ».
+          1. Let _separatorLength_ be the length of _separator_.
+          1. Assert: _separatorLength_ is not 0.
+          1. Let _substrings_ be a new empty List.
+          1. Let _i_ be 0.
+          1. Let _j_ be StringIndexOf(_S_, _separator_, 0).
+          1. Repeat, while _j_ ≠ -1,
+            1. Let _T_ be the substring of _S_ from _i_ to _j_.
+            1. Append _T_ to _substrings_.
+            1. If the number of elements in _substrings_ is _limit_, return _substrings_.
+            1. Set _i_ to _j_ + _separatorLength_.
+            1. Set _j_ to StringIndexOf(_S_, _separator_, _i_).
+          1. Let _T_ be the substring of _S_ from _i_.
+          1. Append _T_ to _substrings_.
+          1. Return _substrings_.
+        </emu-alg>
```

Comment:
> But let's keep it super narrow for now to support evolution of e.g. _limit_ (which really should limit the search for separators rather than truncating the results) and empty string handling.

I'm happy to remove _limit_, but could you explain more about "should limit the search for separators rather than truncating the results"?  Do you mean that _limit_ should cause the last substring to contain any remaining content in the string, even if there are separators in it?

I'm asking because the two cases in the spec so far both perform well using truncation because both of those cases only care about the first few substrings.

Reply:
Also, if we wanted to promote this AO to 262, `String.prototype.split` uses truncation so a "rest" result wouldn't work for that case.

Reply:
> Do you mean that _limit_ should cause the last substring to contain any remaining content in the string, even if there are separators in it?

Yes, that is my position and IIRC there is regret on the committee that `String.prototype.split` itself does not work like that (as it does e.g. [Python `str.split`](https://docs.python.org/3/library/stdtypes.html#str.split)).

> the two cases in the spec so far both perform well using truncation because both of those cases only care about the first few substrings.

I agree, but am not confident that will remain the case and don't want to paint this into a corner. Anything that can be specified with truncation can also be specified without it, and implementations are not obligated to perform unobservable steps (which can even be reiterated in a note if we're concerned about it).

> Also, if we wanted to promote this AO to 262, `String.prototype.split` uses truncation so a "rest" result wouldn't work for that case.

No, it would be fine as e.g.
```
1. Let _substrings_ be StringSplitToList(_S_, _R_, _lim_).
1. If the number of elements in _substrings_ > _lim_, remove the last element from _substrings_.
1. Assert: the number of elements in _substrings_ ≤ _lim_.
1. Return CreateArrayFromList(_substrings_).
```
or even (with a new _maxLength_ parameter for CreateArrayFromList)
```
1. Let _substrings_ be StringSplitToList(_S_, _R_, _lim_).
1. Return CreateArrayFromList(_substrings_, _lim_).
```

Reply:
> Do you mean that _limit_ should cause the last substring to contain any remaining content in the string, even if there are separators in it?

Yes, that is my position and IIRC there is regret on the committee that `String.prototype.split` itself does not work like that (as it does e.g. [Python `str.split`](https://docs.python.org/3/library/stdtypes.html#str.split)).

> the two cases in the spec so far both perform well using truncation because both of those cases only care about the first few substrings.

I agree, but am not confident that will remain the case and don't want to paint this into a corner. Anything that can be specified with truncation can also be specified without it, and implementations are not obligated to perform unobservable steps (which can even be reiterated in a note if we're concerned about it).

> Also, if we wanted to promote this AO to 262, `String.prototype.split` uses truncation so a "rest" result wouldn't work for that case.

No, it would be fine as e.g.
```
1. Let _substrings_ be StringSplitToList(_S_, _R_, _lim_).
1. If the number of elements in _substrings_ > _lim_, remove the last element from _substrings_.
1. Assert: the number of elements in _substrings_ ≤ _lim_.
1. Return CreateArrayFromList(_substrings_).
```
or even (with a new _maxLength_ parameter for CreateArrayFromList)
```
1. Let _substrings_ be StringSplitToList(_S_, _R_, _lim_).
1. Return CreateArrayFromList(_substrings_, _lim_).
```

Reply:
OK thanks for the explanation, I'll remove the limit parameter. Note that the assertion that _S_ isn't empty means that some callsites like `IsWellFormedUnitIdentifier` need to add steps to deal with the empty string case. Is that OK?